Enable more styling options for guide_colorbar().

clauswilke opened the pull request:
Inspired by StackOverflow questions [here](https://stackoverflow.com/q/29793398/4975218) and [here:](https://stackoverflow.com/q/50070741/4975218) This patch enables better styling of the colorbar guide.

Examples:
```{r}
set.seed(123)
df <- data.frame(x <- rnorm(10),
                 y <- rnorm(10),
                 fill <- rnorm(10))

p <- ggplot(df, aes(x, y, fill = fill)) + geom_point()

# current ggplot2
p + scale_fill_gradient(low = 'white', high = 'red')
```
<img width="653" alt="screen shot 2018-04-28 at 10 34 44 am" src="https://user-images.githubusercontent.com/4210929/39398210-6422f5e2-4ad0-11e8-8662-d57048fabbe9.png">

```{r}
# make frame around color bar
p + scale_fill_gradient(low = 'white', high = 'red',
                        guide = guide_colorbar(frame.colour = "black"))
```
<img width="653" alt="screen shot 2018-04-28 at 10 34 54 am" src="https://user-images.githubusercontent.com/4210929/39398217-70c2c34a-4ad0-11e8-991a-37354f1cda64.png">

```{r}
# style tick marks
p + scale_fill_gradient(low = 'white', high = 'red',
                        guide = guide_colorbar(ticks.colour = "black"))
```
<img width="653" alt="screen shot 2018-04-28 at 10 35 01 am" src="https://user-images.githubusercontent.com/4210929/39398221-777fddd0-4ad0-11e8-94db-7e35fdf49954.png">

```{r}
# style frame and tick marks at the same time, change linewidth
p + scale_fill_gradient(low = 'white', high = 'red',
                        guide = guide_colorbar(frame.colour = "black",
                                               frame.linewidth = 1,
                                               ticks.colour = "black",
                                               ticks.linewidth = 1))
```
<img width="653" alt="screen shot 2018-04-28 at 10 35 08 am" src="https://user-images.githubusercontent.com/4210929/39398226-7d496808-4ad0-11e8-8b26-2b1b5c36dcdc.png">

This pull request will require some modifications once https://github.com/tidyverse/ggplot2/pull/2415 is integrated (or vice versa). I'm happy to take care of that.

hadley commented on NEWS.md:
```diff
@@ -285,6 +285,11 @@ up correct aspect ratio, and draws a graticule.
 
+### Guides
+
+* Make `guide_colorbar()` more configurable; enable styling of tick marks and
```
Can you mention the argument names here please?

hadley commented on R/guide-colorbar.r:
```diff
@@ -109,8 +117,15 @@ guide_colourbar <- function(
+  # frame
+  frame.colour = NULL,
+  frame.linewidth = 0.5,
+  frame.linetype = 1,
```
Have we used `linewidth` and `linetype` in other functions?

clauswilke replied:
`linetype` is a standard aesthetic, so I think that's fine. `linewidth` is usually called `size` throughout the rest of ggplot2. If you prefer `size` I'm happy to make the change, it's just that `frame.size` doesn't sound like it controls the thickness of the frame.

hadley replied:
I don't have any particularly strong feelings here; I don't really understand how these options are supposed to fit in with the theme system.

clauswilke replied:
Not sure. The question is whether every possible option to a guide function needs a theme equivalent. My personal opinion is no. I think there is a lot of room for improvement for making guide functions more flexible, and possibly providing different ones or frameworks for extending them, and it's impossible to add new theme components every time somebody makes some minor tweak to some guide function. Already, there are things the guide functions can do that the theme can't control, such as the `draw.ulim` option for `guide_colorbar()`.

hadley replied:
Agreed. I'm happy to merge this PR; I just wanted to make sure we at least talked over the issues.

hadley commented on R/guide-colorbar.r:
```diff
@@ -273,6 +295,16 @@ guide_gengrob.colorbar <- function(guide, theme) {
+
+  # make frame around color bar if requested (colour is not NULL)
+  if (!is.null(guide$frame.colour)) {
+    grob.bar <- grobTree(grob.bar,
```
Would you mind adhering to modern tidyverse style here?

```R

grobTree(
  grob.bar,
  rectGrob(
    width = barwidth.c, height = barheight.c, default.units = "mm",
    gp = gpar(
      col = guide$frame.colour,
      lwd = guide$frame.linewidth,
      lty = guide$frame.linetype,
      fill = NA
    )
  )
)
```

or similar

clauswilke replied:
Sure, will do.

hadley commented on R/guide-colorbar.r:
```diff
@@ -352,7 +384,9 @@ guide_gengrob.colorbar <- function(guide, theme) {
         })
       segmentsGrob(x0 = x0, y0 = y0, x1 = x1, y1 = y1,
-                   default.units = "mm", gp = gpar(col = "white", lwd = 0.5, lineend = "butt"))
+                   default.units = "mm", gp = gpar(col = guide$ticks.colour,
```
And here

clauswilke commented:
Just a note: I'll wait with finalizing this PR until after https://github.com/tidyverse/ggplot2/pull/2415 is integrated, since they touch the same files.

clauswilke commented:
I believe this latest update addresses all concerns. I added two minimal visual tests, one for the default styling (it also tests that colors can be changed via scale_color_..., so is useful anyways I think) and one for the modified styling of both the frame and the tick lines.

hadley commented:
Thank you!
